modal/cli/run.py

```diff
@@ -368,24 +369,30 @@ def shell(
     if not console.is_terminal:
         raise click.UsageError("`modal shell` can only be run from a terminal.")
 
+    stub = Stub("modal shell")
+
     if func_ref is not None:
-        function = import_function(func_ref, accept_local_entrypoint=False, accept_webhook=True, base_cmd="modal shell")
+        function: Function = import_function(
+            func_ref, accept_local_entrypoint=False, accept_webhook=True, base_cmd="modal shell"
+        )
+        function_env: FunctionEnv = function.env
+        interactive_shell(
+            stub,
+            cmd,
+            environment_name=env,
+            **asdict(function_env),
+            timeout=3600,
+        )
     else:
-        image_obj = Image.from_registry(image, add_python=add_python) if image else None
-        stub = Stub("modal shell", image=image_obj)
-        function = stub.function(
-            serialized=True,
+        image = Image.from_registry(image, add_python=add_python) if image else None
+        interactive_shell(
```

Comment:
Minor but would be cleaner to have the if/else just set the function environment, and then just one call to `interactive_shell`.